### PR TITLE
Merge Segments for line detection

### PR DESCRIPTION
**Pull request by knoellle**
## Introduced Changes

What it says. We originally were merging consecutive non-field segments directly in the segment filter but realized that feet detection and ball detection rely on counting segments.
Now the merging happens on the fly when generating seed points for ransac in the line detection node.

This PR is part of a set:
- #683
- #684
- #686
- #687

 ~~Collect them all for a chance to win a golden PR! Buy now and win!~~
If you wish to test them all together, use this branch: [fliewatüüt](https://github.com/knoellle/hulk/tree/fliewat%C3%BC%C3%BCt)

## Main

image

## This branch

image

Fixes #

## Ideas for Next Iterations (Not This PR)

Investigate merging segments based on distance in robot coordinates instead of distance in pixels.

## How to Test

Have a look at images with wide (bottom camera), noisy (sprayed for example) lines.
On main, there would be many small ransac seed points all over the place but on this branch they should be a lot closer to the center of the line.

### CODE REVIEW

**Review by schluis — CHANGES_REQUESTED** [review]
Very nice.

Since you're editing this file, do you think it's feasable to use the new `CoordinateSystems` here?

File: etc/parameters/default.json [hunk 1]
```diff
@@ -119,7 +120,8 @@
       "maximum_gap_on_line": 30.0,
       "maximum_number_of_lines": 10,
       "maximum_projected_segment_length": 0.3,
-      "minimum_number_of_points_on_line": 4
+      "minimum_number_of_points_on_line": 4,
+      "maximum_merge_gap": 30
```
[the remark this answers is not in the document]
Comment (knoellle) [reply etc/parameters/default.json]: Done

File: crates/vision/src/line_detection.rs [hunk 2]
```diff
@@ -234,6 +237,47 @@ fn get_gradient(image: &YCbCr422Image, point: Point2<u16>) -> Vector2<f32> {
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut current = self.iterator.next()?;
+
+        while let Some(next) = self.iterator.peek().copied() {
+            if next.start - current.end >= self.maximum_merge_gap {
+                break;
+            }
+
+            let _ = self.iterator.next();
```
Comment (schluis) [inline crates/vision/src/line_detection.rs:271]: I think you should be able to remove `let _ =`

File: crates/vision/src/line_detection.rs [hunk 3]
```diff
@@ -331,110 +379,9 @@ fn is_segment_shorter_than(
...
-            &filtered_segments,
-            &image,
-            check_line_segments_projection,
-            maximum_projected_segment_length,
-            check_edge_gradient,
-            gradient_alignment,
-        );
-        assert_eq!(line_points.len(), 32);
-    }
-
```
Comment (schluis) [inline crates/vision/src/line_detection.rs:437]: Why did you remove the tests? Because they failed? 😆

**Review by schluis — CHANGES_REQUESTED** [review]

File: crates/vision/src/line_detection.rs (hunk 2 above)
Comment (knoellle) [reply crates/vision/src/line_detection.rs:271]: Yes, but I think this is more readable. I explicitly and intentionally discard the return value here.

File: crates/vision/src/line_detection.rs (hunk 3 above)
Comment (schluis) [reply crates/vision/src/line_detection.rs:437]: Okay, please elaborate...

Comment (knoellle) [reply crates/vision/src/line_detection.rs:437]: Since we were previously applying the median in the same direction we were constructing segments, we were actually using a sort of moving median window.
However, since the median direction and segment direction are now orthogonal, every pixel is only inspected once (but more pixels are looked at in total) now.

Yes, we can construct new test data which would make these tests "work" again, but we already have tests without median and the medians are tested separately anyways.
Given that the two are orthogonal now, I see little value in trying to repair these tests.

Comment (schluis) [reply crates/vision/src/line_detection.rs:437]: Okay, makes sence 👍🏻

**Review by schluis — APPROVED** [review]
